### ISSUE TITLE
Test failure in macros::args on beta Rust

### ISSUE THREAD

**Issue by djc** [djc/askama#504]
@vallentin did you already have some attempt to minimize the reproduction for this? I really want to report it upstream soon so we can understand what's going on here.

**Comment by vallentin** [comment]
Yes, but I'm not home until tomorrow :)

**Comment by djc** [comment]
Ok.

**Comment by djc** [comment]
@vallentin friendly ping?

**Comment by Kijewski** [comment]
The diff between stable and testing is:

```diff
-"some ninety \"(\\\"()\"nine !"
+"some ninety \"(\\\"()\"nine!"
```

I.e. stable adds a spaces in front of the bang. Since `call_a_or_b_on_tail!()` operates on a TokenTree, I guess it is allowed to add extraneous spaces, or omit them if they are not needed. IMO the old behavior was wrong, but rust_macro.rs operates under the assumption that the space gets added implicitly.

**Comment by djc** [comment]
Thanks for looking into this! Do you have any suggestions on how to make rust_macro.rs robust to these kinds of changes?

**Comment by vallentin** [comment]
@Kijewski already identified the issue and posted it on Gitter, but thanks anyways :)

Having `stringify!(foo ! bar !)` snap exclamation marks to the left _does make sense_ to a degree, if and only if the goal of `stringify!` is to loosely mirror rustfmt. So I guess the question is more whether stringification should take everything literal, or if it should _format_ it, which I don't think it should, given it's called `stringify` and not `rustfmtify`.

I'm looking for the commit and PR where this was changed, to see if it was acknowledged or an accidental side-effect of another change.

-----

In regards to testing, then I noticed that the [`stringify!`](https://doc.rust-lang.org/std/macro.stringify.html) docs, includes the following:

> Note that the expanded results of the input tokens may change in the future. You should be careful if you rely on the output.

So I guess we should be wary of using `stringify!` in tests.

**Comment by Kijewski** [comment]
No, not really. I guess the only option is to write a test that does not depend on the output of stringify.

**Comment by djc** [comment]
Ah, the offending commit is https://github.com/djc/askama/commit/39e08325c0e0fba75384579427d1c6f7eca8fbc9, that's much older than I thought.

**Comment by vallentin** [comment]
Yes, that the test here in askama that's the issue.

Though to be clear, I'm looking for the commit and PR in the rust repo ofc :)

**Comment by vallentin** [comment]
@djc should we remove the test or change it to not use `stringify!`? I think it's probably best to avoid `stringify!` in tests based on the line in the docs

**Comment by djc** [comment]
If we can rewrite the test without using stringify, that would be great!

**Comment by djc** [comment]
Fixed by #511.
